lib/streamlit/dataframe_util.py

```diff
@@ -334,64 +332,59 @@ def convert_anything_to_pandas_df(
         ) from ex
 
 
-@overload
-def ensure_iterable(obj: Iterable[V_co]) -> Iterable[V_co]: ...
+def convert_anything_to_sequence(obj: OptionSequence[V_co]) -> Sequence[V_co]:
+    """Try to convert different formats to an indexable Sequence.
 
-
-@overload
-def ensure_iterable(obj: OptionSequence[V_co]) -> Iterable[Any]: ...
-
-
-def ensure_iterable(obj: OptionSequence[V_co] | Iterable[V_co]) -> Iterable[Any]:
-    """Try to convert different formats to something iterable. Most inputs
-    are assumed to be iterable, but if we have a DataFrame, we can just
-    select the first column to iterate over. If the input is not iterable,
+    If the input is a dataframe-like object, we can just select the first
+    column to iterate over. If the input cannot be converted to a sequence,
     a TypeError is raised.
 
     Parameters
     ----------
-    obj : list, tuple, numpy.ndarray, pandas.Series, pandas.DataFrame, pyspark.sql.DataFrame, snowflake.snowpark.dataframe.DataFrame or snowflake.snowpark.table.Table
+    obj : OptionSequence
+        The object to convert to a sequence.
 
     Returns
     -------
-    iterable
-
+    Sequence
+        The converted sequence.
     """
+    if obj is None:
+        return []  # type: ignore
+
+    if isinstance(obj, str):
+        # String is a special case: theoretically it is iterable, but we don't want
+        # to iterate over its characters. Instead, we raise a TypeError.
+        raise TypeError(
+            "Object is not an iterable and could not be converted to one. "
+            f"Object type: {type(obj)}"
+        )
```

Comment:
Hm, I'm not 100% sure we're able to do this since there may be some rare cases where people are passing strings to options widgets to abuse the fact that they're iterable. Certainly a weird case, but this will be a breaking change for the people that do happen to be doing this.

Reply:
Oh, interesting. I actually changed this since using string with multiselect doesn't work, and we have a test checking that here:

https://github.com/streamlit/streamlit/blob/b4681fb40788f5860f8fe21a5966d092d0acd52d/lib/tests/streamlit/elements/multiselect_test.py#L138-L139

But string in `radio`, `select_slider` and `selectbox` does work, so `multiselect` seems to be the outlier. I will change it back to support string and also fix this in multiselect.